sdk/core/System.TypeSpec/src/PipelineOptions.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ComponentModel;
+using System.Diagnostics;
+using System.ServiceModel.Rest.Core;
+using System.Threading;
+
+namespace System.ServiceModel.Rest;
+
+// Make options freezable
+public class PipelineOptions
+{
+    private object _lock = new object();
+    private MessagePipeline? _pipeline;
+    private volatile bool _isFrozen;
+
+    private IPipelinePolicy<PipelineMessage>? _retryPolicy;
+    private IPipelinePolicy<PipelineMessage>? _loggingPolicy;
+    private PipelineTransport<PipelineMessage> _transport;
+    private ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> _perTryPolicies;
+    private ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> _perCallPolicies;
+
+    public PipelineOptions(PipelineTransport<PipelineMessage> transport)
+    {
+        _transport = transport;
+    }
+
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public virtual MessagePipeline GetPipeline()
+    {
+        if (_pipeline != null) return _pipeline;
+
+        lock (_lock) {
+            if (IsFrozen) {
+                Debug.Assert(_pipeline != null);
+                return _pipeline!;
+            }
+            _pipeline = MessagePipeline.Create(this);
+            Freeze();
+            return _pipeline;
+        }
+    }
+
+    protected bool IsFrozen => _isFrozen;
+    protected void Freeze() => _isFrozen = true;
```

Comment:
Why do these need to be protected?  Should they be visible in the public API?  i.e. could they be internal virtual only so they're not visible outside the assembly?

Reply:
I think I was envisioning options freezing themselves when a property is read, i.e. so no need for public API.

Reply:
I think it can be private, then, and I'll do that for now until we learn otherwise.